File: frontend/src/otelConfig.ts
/**
 * OpenTelemetry Web SDK configuration for TransKeep frontend
 * 
 * Initializes distributed tracing for the React application,
 * sending traces to Jaeger collector for development.
 */

import { BasicTracerProvider, SimpleSpanProcessor } from '@opentelemetry/sdk-trace-web'
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http'
import { Resource } from '@opentelemetry/resources'
import { ATTR_SERVICE_NAME, ATTR_SERVICE_VERSION } from '@opentelemetry/semantic-conventions'
import { trace } from '@opentelemetry/api'

/**
 * Initialize OpenTelemetry Web SDK with Jaeger exporter
 * 
 * @param serviceName - Name of the service for trace identification
 * @param collectorUrl - OTLP HTTP collector endpoint URL
 */
export function initTelemetry(
  serviceName: string = 'transkeep-frontend',
  collectorUrl: string = 'http://localhost:4318/v1/traces'
): void {
  // Create OTLP exporter for sending traces
  const exporter = new OTLPTraceExporter({
    url: collectorUrl,
  })

  // Create resource with service name and metadata
  const resource = Resource.default().merge(
    new Resource({
      [ATTR_SERVICE_NAME]: serviceName,
      [ATTR_SERVICE_VERSION]: '0.1.0',
    })
  )

  // Create tracer provider with resource
  const tracerProvider = new BasicTracerProvider({
    resource: resource,
  })

  // Add exporter as span processor
  tracerProvider.addSpanProcessor(new SimpleSpanProcessor(exporter))

  // Set global tracer provider
  tracerProvider.register()

  console.log(`✓ OpenTelemetry initialized for ${serviceName}`)
}

/**
 * Get current tracer for creating custom spans
 * 
 * @param name - Module name or identifier for the tracer
 * @returns Tracer instance for custom instrumentation
 */
export function getTracer(name: string = 'transkeep-app') {
  return trace.getTracer(name)
}

/**
 * Get trace ID from current context (for logging correlation)
 * 
 * @returns Current trace ID as hex string, or undefined if not in trace
 */
export function getCurrentTraceId(): string | undefined {
  // This would be populated by the tracer when in an active span
  // For now, return undefined - will be enhanced with context API
  return undefined
}

File: backend/app/otel_config.py
"""OpenTelemetry configuration for TransKeep backend"""

import os
from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.exporter.jaeger.thrift import JaegerExporter
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.instrumentation.sqlalchemy import SQLAlchemyInstrumentor
from opentelemetry.instrumentation.requests import RequestsInstrumentor
from opentelemetry.sdk.resources import SERVICE_NAME, Resource


def init_telemetry(service_name: str = "transkeep-backend") -> TracerProvider:
    """
    Initialize OpenTelemetry instrumentation with Jaeger exporter.
    
    Args:
        service_name: Name of the service for trace identification
    
    Returns:
        TracerProvider instance
    """
    # Get Jaeger configuration from environment
    jaeger_host = os.getenv("OTEL_EXPORTER_JAEGER_AGENT_HOST", "localhost")
    jaeger_port = int(os.getenv("OTEL_EXPORTER_JAEGER_AGENT_PORT", 6831))
    
    # Create Jaeger exporter
    jaeger_exporter = JaegerExporter(
        agent_host_name=jaeger_host,
        agent_port=jaeger_port,
    )
    
    # Create resource with service name
    resource = Resource(attributes={
        SERVICE_NAME: service_name
    })
    
    # Create tracer provider with resource
    tracer_provider = TracerProvider(resource=resource)
    
    # Add Jaeger exporter as batch processor
    tracer_provider.add_span_processor(
        BatchSpanProcessor(jaeger_exporter)
    )
    
    # Set the global tracer provider
    trace.set_tracer_provider(tracer_provider)
    
    return tracer_provider


def instrument_app(app) -> None:
    """
    Instrument FastAPI application and common libraries.
    
    Args:
        app: FastAPI application instance
    """
    # Instrument FastAPI endpoints
    FastAPIInstrumentor.instrument_app(app)
    
    # Instrument SQLAlchemy database calls
    SQLAlchemyInstrumentor().instrument()
    
    # Instrument HTTP requests
    RequestsInstrumentor().instrument()


def get_tracer(name: str = "transkeep-backend") -> trace.Tracer:
    """
    Get a tracer instance for custom instrumentation.
    
    Args:
        name: Module name or identifier for the tracer
    
    Returns:
        Tracer instance for creating custom spans
    """
    return trace.get_tracer(name)

